README
NAME

    Bit::Manip - Functions to simplify bit string manipulation

SYNOPSIS

        use Bit::Manip qw(:all);
    
        my $b;    # bit string
        $b = 128; # 10000000
    
        $b = bit_toggle($b, 4); # 10010000
        $b = bit_toggle($b, 4); # 10000000
        
        $b = bit_off($b, 7);    # 0 
        $b = bit_on($b, 7);     # 10000000 
    
        # get the value of a range of bits...
        # in this case, we'll print the value of bits 4-3
    
        $b = 0b00111000; (56)
    
        print bit_get($b, 4, 3); # 3
    
        # set a range of bits...
        # let's set bits 4-2 to binary 101
    
        $b = 0b10000000;
        my $num_bits = 3; # 0b101 in the call is 3 bits
    
        $b = bit_set($b, 2, $num_bits, 0b101); # 10010100
    
        # clear some bits
    
        $b = 0b11111111;
    
        $num_bits = 3;
        $lsb = 3;
    
        $b = bit_clr($b, $lsb, $num_bits); # 11000111
    
        # helpers
    
        my ($num_bits, $lsb) = (3, 2);
    
        print bit_mask($num_bits, $lsb); # 28, or 11100
    
        print bit_bin(255); # 11111111 (same as printf("%b", 255);)
          

DESCRIPTION

    Provides functions to aid in bit manipulation (set, unset, toggle,
    shifting) etc. Particularly useful for embedded programming and writing
    device communication software.

    Currently, up to 32-bit integers are supported.

EXPORT_OK

    Use the :all tag (eg: use Bit::Manip qw(:all);) to import the following
    functions into your namespace, or pick and choose individually:

        bit_bin
        bit_count 
        bit_mask
        bit_get 
        bit_set 
        bit_toggle
        bit_on
        bit_off

FUNCTIONS

 bit_bin

    Returns the binary representation of a number as a string of ones and
    zeroes.

    Parameters:

        $data

    Mandatory: Integer, the number you want to convert.

 bit_count

    Returns either the total count of bits in a number, or just the number
    of set bits (if the $set, parameter is sent in and is true).

    Parameters:

        $num

    Mandatory: Unsigned integer, the number to retrieve the total number of
    bits for. For example, if you send in 15, the total number of bits
    would be 4, likewise, for 255, the number of bits would be 16.

        $set

    Optional: Integer. If this is sent and is a true value, we'll return
    the number of *set* bits only. For example, for 255, the set bits will
    be 8 (ie. all of them), and for 8, the return will be 1 (as only the
    MSB is set out of all four of the total).

    Return: Integer, the number of bits that make up the number if $set is
    0, and the number of set bits (1's) if $set is true.

 bit_get

    Retrieves the value of specified bits within a bit string.

    Parameters:

        $data

    Mandatory: Integer, the bit string you want to send in. Eg: 255 for
    11111111 (or 0xFF).

        $msb

    Mandatory: Integer, the Most Significant Bit (leftmost) of the group of
    bits to collect the value for (starting from 0 from the right, so with
    1000, so you'd send in 3 as the start parameter for the bit set to 1).
    Must be 1

        $lsb

    Optional: Integer, the Least Significant Bit (rightmost) of the group
    of bits to collect the value for (starting at 0 from the right). A
    value of 0 means return the value from $msb through to the very end of
    the bit string. A value of 1 will capture from $msb through to bit 1
    (second from right). This value must be lower than $msb.

    Return: Integer, the modified $data param.

 bit_set

    Allows you to set a value for specific bits in your bit string.

    Parameters:

        $data

    Mandatory: Integer, the bit string you want to manipulate bits in.

        $lsb

    Mandatory: Integer, the least significant bit (rightmost) in the bit
    range you want to manipulate. For example, if you wanted to set a new
    value for bits 7-5, you'd send in 5.

        $nbits

    Mandatory: Integer, the number of bits you're sending in. We need this
    param in the event your leading bit is a zero. For example, if you're
    sending in 0b111 or 0b001, this param would be 3.

        $value

    Mandatory: Integer, the value that you want to change the specified
    bits to. Easiest if you send in a binary string (eg: 0b1011 in Perl).

    Return: Integer, the modified $data param.

    Example:

    You have an 8-bit register where the MSB is a start bit, and the rest
    of the bits are zeroed out:

        my $data = 0b10000000; # (0x80, or 128)

    The datasheet for the hardware you're writing to requires you to set
    bits 6-4 to 111 in binary (always start from bit 0, not 1):

        10000000
         ^^^   ^
         6-4   0

    Code:

        my $x = bit_set($data, 4, 3, 0b111); # (0x07, or 7)
        printf("%b\n", $x); # prints 11110000

 bit_clr

    Clear (unset to 0) specific bits in the bit string.

    Parameters:

        $data

    Mandatory: Integer, the bit string you want to manipulate bits in.

        $lsb

    Mandatory: Integer, the least significant bit (rightmost) in the bit
    range you want to manipulate. For example, if you wanted to clear bits
    7-5, you'd send in 5.

        $nbits

    Mandatory: Integer, the number of bits you're wanting to clear,
    starting from the $lsb bit, and clearing the number of bits to the
    left.

    Returns the modified bit string.

 bit_toggle

    Toggles a single bit. If it's 0 it'll toggle to 1 and vice-versa.

    Parameters:

        $data

    Mandatory: Integer, the number/bit string to toggle a bit in.

        $bit

    Mandatory: Integer, the bit number counting from the right-most (LSB)
    bit starting from 0.

    Return: Integer, the modified $data param.

 bit_on

    Sets a single bit (sets to 1), regardless of its current state.

    Parameters:

        $data

    Mandatory: Integer, the number/bit string to toggle a bit in.

        $bit

    Mandatory: Integer, the bit number counting from the right-most (LSB)
    bit starting from 0.

    Return: Integer, the modified $data param.

 bit_off

    Unsets a single bit (sets to 0), regardless of its current state.

    Parameters:

        $data

    Mandatory: Integer, the number/bit string to toggle a bit in.

        $bit

    Mandatory: Integer, the bit number counting from the right-most (LSB)
    bit starting from 0.

    Return: Integer, the modified $data param.

 bit_mask

    Generates a bit mask for the specific bits you specify.

    Parameters:

        $nbits

    Mandatory: Integer, the number of bits to get the mask for.

        $lsb

    Mandatory: Integer, the LSB at which you plan on implementing your
    change.

    Return: Integer, the bit mask ready to be applied.

AUTHOR

    Steve Bertrand, <steveb at cpan.org>

LICENSE AND COPYRIGHT

    Copyright 2017 Steve Bertrand.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Bit-Manip

1.05 UNREL
        -

1.04    2017-07-23
        - replaced all calls to pow() with bit shift operations instead,
          breaking reliance on stdint.h (closes #10)
        - renamed bit_toggle() to bit_tog(), left an alias for backward
          compatibility purposes (closes #8)
        - add the parameter lists to the functions in POD (closes #9)
        - changed to the older way of using Exporter (closes #7)
        - added note in DESCRIPTION that we also have a pure Perl version
          (closes #6)
        - verified $msb can equal $lsb in bit_get(), and added tests to ensure
          it (closes #2)
        - added tests to ensure bit_tog() does the right thing

1.03    2017-02-09
        - added bit_clr() (per PerlMonks discussion), clears selected bits
        - changed $bits param name to $nbits in code and POD as it makes it less
          ambiguous as to what we mean by the arg
        - added t/45 for testing bit_clr()

1.02    2016-01-27
        - removed erroneous "use feature 'say';" from test files, as it was
          causing Testers to break on versions < v5.10.x

1.01    2016-01-25
        - somehow I missed adding POD for the changes in bit_set()

1.00    2016-01-25
        - removed string.h include as we don't use it
        - LSB can now be equal to MSB to allow getting a single bit
        - fix issue where bit_set() was not updating bits if a leading zero was
          sent in. We now require a 'bits' param (fixes #1)

0.02    2017-01-25
        - added in EXPORT_OK POD section
        - code tidying in the XS file
        - changed all instances of "first" to msb and "last" to lsb for
          params
        - added bit_mask() returns the bit mask given a number of bits and LSB

0.01    2017-01-24
        First version, released on an unsuspecting world.



examples/test.pl
use warnings;
use strict;
use feature 'say';

use Bit::Manip::PP qw(:all);

my $data = bit_on(0, 15);

$data = bit_set($data, 13, 0b11);
say bit_bin($data);

$data = bit_set($data, 11, 0b01);
say bit_bin($data);

$data = bit_set($data, 0, 186);
say bit_bin($data);

$data = bit_set($data, 13, 0b01);
say bit_bin($data);

$data = bit_off($data, 14);
say bit_bin($data);

say "break";

say bit_bin(bit_get($data, 10, 8));




bit.pl
use warnings;
use strict;
use feature 'say';

use Bit::Manip qw(:all);
use Test::More;

is bin(bit_set(8, 0, 0b1)), '1001', "8, 0, 1 ok";
is bin(bit_set(8, 1, 0b1)), '1010', "8, 1, 1 ok";
is bin(bit_set(8, 2, 0b1)), '1100', "8, 2, 1 ok";

is bin(bit_set(8, 0, 0b111)), '1111', "8, 0, 0b111 ok";
is bin(bit_set(8, 1, 0b011)), '1110', "8, 1, 0b11 ok";
is bin(bit_set(8, 2, 0b1)), '1100', "8, 2, 0b01 ok";

sub bin {
    return sprintf "%b", $_[0];
}

done_testing();

__END__
$x = bit_set(128, 0, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 1, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 2, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 3, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 4, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 5, 1);
printf("%d: %b\n", $x, $x);


say "bit 2";

$x = bit_set(128, 2, 0);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 2, 1);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 2, 2);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 2, 3);
printf("%d: %b\n", $x, $x);

say "bit 3";

$x = bit_set(128, 3, 0b11);
printf("%d: %b\n", $x, $x);

$x = bit_set(128, 3, 0b111);
printf("%d: %b\n", $x, $x);

$x = bit_set(255, 0, 0b0);
printf("%d: %b\n", $x, $x);

__END__

is bit_set(128, 0, 1), 129, "128, 0, 1 ok";
is bit_set(2, 0, 1), 3, "2, 0, 1 ok";

is bit_set(0, 5, 0b10), 256, "255, 0, 1 ok";

done_testing();


examples/seq_skip_prototype.pl
use warnings;
use strict;
use feature 'say';

use Bit::Manip qw(:all);
use Test::More;

# out of order sequence

for (0..11){
    bit_set_seq(0x00, [3, 1, 2, 0], $_);
}

# skip over

for (0..11){
    my $b = $_;
    $b = bit_on($b, 1);
    $b = bit_on($b, 2);
    bit_set_skip($b, [0, 3], $_);
}

sub bit_set_seq {
    my ($b, $seq, $val) = @_;

    my $lsb = (sort @$seq)[0];
    my $msb = (sort @$seq)[-1];

    my $nbits = ($msb - $lsb);

    $b = bit_set($b, $lsb, $nbits, $val);

    my $x = $b;

    for (0..$nbits){
        my $o = bit_get($b, $_, $_);
        my $s = bit_get($b, $seq->[$_], $seq->[$_]);
        
        if ($o != $s){
            $x = bit_set($x, $_, 1, $s);
        }     
    }
}
sub bit_set_skip {
    my ($b, $seq, $val) = @_;

    my $lsb = (sort @$seq)[0];
    my $msb = (sort @$seq)[-1];

    my @leave_bits 
      = map $seq->[$_-1]+1..$seq->[$_]-1, 1..@$seq-1;

    my %leave_vals 
      = map {$_ => bit_get($b, $_, $_)} @leave_bits;

    my $nbits = ($msb - $lsb) + 1;

    $b = bit_set($b, $lsb, $nbits, $val);

    say "** $val";

    printf("skip before: %b\n", $b);

    for (keys %leave_vals){
        $b = bit_set($b, $_, 1, $leave_vals{$_});
    }

    printf("skip after:  %b\n\n", $b);
}


lib/Bit/Manip.pm
package Bit::Manip;

use warnings;
use strict;

our $VERSION = '1.05';

require XSLoader;
XSLoader::load('Bit::Manip', $VERSION);

use Exporter;
our @ISA = qw(Exporter);

our @EXPORT_OK = qw(
    bit_get
    bit_set
    bit_clr
    bit_bin
    bit_count
    bit_mask
    bit_tog
    bit_toggle
    bit_on
    bit_off
);

our %EXPORT_TAGS;
$EXPORT_TAGS{all} = [@EXPORT_OK];

sub bit_bin {
    my ($data) = @_;
    return sprintf("%b", $data);
}
sub bit_count {
    my ($n, $set) = @_;

    if (! defined $n || $n !~ /^\d+/){
        die "bit_count() requires an integer param\n";
    }

    $set = 0 if ! defined $set;

    return _bit_count($n, $set);
}
sub bit_mask {
    my ($nbits, $lsb) = @_;
    return _bit_mask($nbits, $lsb);
}
sub bit_get {
    my ($data, $msb, $lsb) = @_;

    $lsb = 0 if ! defined $lsb;

    _bit_get($data, $msb, $lsb);
}
sub bit_clr {
    my ($data, $lsb, $nbits) = @_;
    return _bit_set($data, $lsb, $nbits, 0);
}
sub bit_set {
    my ($data, $lsb, $nbits, $value) = @_;
    return _bit_set($data, $lsb, $nbits, $value);
}

# alias bit_toggle() into play

*bit_toggle = \&bit_tog;

sub bit_tog {
    my ($data, $bit) = @_;
    return _bit_tog($data, $bit);
}
sub bit_on {
    my ($data, $bit) = @_;
    return _bit_on($data, $bit);
}
sub bit_off {
    my ($data, $bit) = @_;
    return _bit_off($data, $bit);
}
sub _vim{};

1;
__END__

=head1 NAME

Bit::Manip - Functions to simplify bit string manipulation

=head1 SYNOPSIS

    use Bit::Manip qw(:all);

    my $b;    # bit string
    $b = 128; # 10000000

    # toggle a bit off and on

    $b = bit_tog($b, 4); # 10010000
    $b = bit_tog($b, 4); # 10000000

    # turn a bit off, then back on

    $b = bit_off($b, 7);    # 0 
    $b = bit_on($b, 7);     # 10000000 

    # get the value of a range of bits...
    # in this case, we'll print the value of bits 4-3

    $b = 0b00111000; (56)

    print bit_get($b, 4, 3); # 3

    # set a range of bits...
    # let's set bits 4-2 to binary 101

    $b = 0b10000000;
    my $num_bits = 3; # 0b101 in the call is 3 bits

    $b = bit_set($b, 2, $num_bits, 0b101); # 10010100

    # clear some bits

    $b = 0b11111111;

    $num_bits = 3;
    $lsb = 3;

    $b = bit_clr($b, $lsb, $num_bits); # 11000111

    # helpers

    my ($num_bits, $lsb) = (3, 2);
    print bit_mask($num_bits, $lsb); # 28, or 11100

    print bit_bin(255); # 11111111 (same as printf("%b", 255);)
      
=head1 DESCRIPTION

Provides functions to aid in bit manipulation (get, set, clear, toggle)
etc. Particularly useful for embedded programming and writing device
communication software.

Currently, up to 32-bit integers are supported.

Note that if your machine isn't capable of compiling C code, there's a pure Perl
version as well, aptly named L<Bit::Manip::PP>.

=head1 EXPORT_OK

Use the C<:all> tag (eg: C<use Bit::Manip qw(:all);>) to import the following
functions into your namespace, or pick and choose individually:

    bit_get
    bit_set
    bit_clr
    bit_tog
    bit_on
    bit_off
    bit_bin
    bit_count
    bit_mask

=head1 FUNCTIONS

=head2 bit_get($data, $msb, $lsb)

Retrieves the value of specified bits within a bit string.

Parameters:

    $data

Mandatory: Integer, the bit string you want to send in. Eg: C<255> for
C<11111111> (or C<0xFF>).

    $msb

Mandatory: Integer, the Most Significant Bit (leftmost) of the group of bits to
collect the value for (starting from 0 from the right, so with C<1000>, so you'd
send in C<3> as the start parameter for the bit set to C<1>). Must be C<1>

    $lsb

Optional: Integer, the Least Significant Bit (rightmost) of the group of bits to
collect the value for (starting at 0 from the right). A value of C<0> means
return the value from C<$msb> through to the very end of the bit string. A
value of C<1> will capture from C<$msb> through to bit C<1> (second from
right). If C<$msb> is equal to C<$lsb>, we'll return just that bit.

Return: Integer, the modified C<$data> param.

=head2 bit_set($data, $lsb, $nbits, $value)

Allows you to set a value for specific bits in your bit string.

Parameters:

    $data

Mandatory: Integer, the bit string you want to manipulate bits in.

    $lsb

Mandatory: Integer, the least significant bit (rightmost) in the bit range you
want to manipulate. For example, if you wanted to set a new value for bits
C<7-5>, you'd send in C<5>.

    $nbits

Mandatory: Integer, the number of bits you're sending in. We need this param
in the event your leading bit is a zero. For example, if you're sending in
C<0b111> or C<0b001>, this param would be C<3>.

    $value

Mandatory: Integer, the value that you want to change the specified bits to.
Easiest if you send in a binary string (eg: C<0b1011> in Perl).

Return: Integer, the modified C<$data> param.

Example: 

You have an 8-bit register where the MSB is a start bit, and the rest
of the bits are zeroed out:

    my $data = 0b10000000; # (0x80, or 128)

The datasheet for the hardware you're writing to requires you to set bits 
C<6-4> to C<111> in binary (always start from bit 0, not 1):

    10000000
     ^^^   ^
     6-4   0

Code:

    my $x = bit_set($data, 4, 3, 0b111); # (0x07, or 7)
    printf("%b\n", $x); # prints 11110000

=head2 bit_clr($data, $lsb, $nbits)

Clear (unset to 0) specific bits in the bit string.

Parameters:

    $data

Mandatory: Integer, the bit string you want to manipulate bits in.

    $lsb

Mandatory: Integer, the least significant bit (rightmost) in the bit range you
want to manipulate. For example, if you wanted to clear bits C<7-5>, you'd send
in C<5>.

    $nbits

Mandatory: Integer, the number of bits you're wanting to clear, starting from
the C<$lsb> bit, and clearing the number of bits to the left.

Returns the modified bit string.

=head2 bit_toggle($data, $bit)

See L</bit_tog>.

=head2 bit_tog($data, $bit)

AKA: C<bit_toggle()>.

Toggles a single bit. If it's C<0> it'll toggle to C<1> and vice-versa.

Parameters:

    $data

Mandatory: Integer, the number/bit string to toggle a bit in.

    $bit

Mandatory: Integer, the bit number counting from the right-most (LSB) bit
starting from C<0>.

Return: Integer, the modified C<$data> param.

=head2 bit_on($data, $bit)

Sets a single bit (sets to C<1>), regardless of its current state. This is just
a short form of setting a single bit with L<bit_set>.

Parameters:

    $data

Mandatory: Integer, the number/bit string to toggle a bit in.

    $bit

Mandatory: Integer, the bit number counting from the right-most (LSB) bit
starting from C<0>.

Return: Integer, the modified C<$data> param.

=head2 bit_off($data, $bit)

Unsets a single bit (sets to C<0>), regardless of its current state. This is
just a short form of clearing a single bit with L<bit_set>.

Parameters:

    $data

Mandatory: Integer, the number/bit string to toggle a bit in.

    $bit

Mandatory: Integer, the bit number counting from the right-most (LSB) bit
starting from C<0>.

Return: Integer, the modified C<$data> param.

=head2 bit_bin($data)

Returns the binary representation of a number as a string of ones and zeroes.

Parameters:

    $data

Mandatory: Integer, the number you want to convert.

=head2 bit_count($num, $set)

Returns either the total count of bits in a number, or just the number of set
bits (if the C<$set>, parameter is sent in and is true).

Parameters:

    $num

Mandatory: Unsigned integer, the number to retrieve the total number of bits
for. For example, if you send in C<15>, the total number of bits would be C<4>,
likewise, for C<255>, the number of bits would be C<16>.

    $set

Optional: Integer. If this is sent and is a true value, we'll return the number
of *set* bits only. For example, for C<255>, the set bits will be C<8> (ie. all
of them), and for C<8>, the return will be C<1> (as only the MSB is set out of
all four of the total).

Return: Integer, the number of bits that make up the number if C<$set> is C<0>,
and the number of set bits (1's) if C<$set> is true.

=head2 bit_mask($nbits, $lsb)

Generates a bit mask for the specific bits you specify.

Parameters:

    $nbits

Mandatory: Integer, the number of bits to get the mask for.

    $lsb

Mandatory: Integer, the LSB at which you plan on implementing your change.

Return: Integer, the bit mask ready to be applied.

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

